File: Python/110.平衡二叉树.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if self.maxDepth(root) == -1:
            return False
        return True

    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        left: int = self.maxDepth(root.left)
        right: int = self.maxDepth(root.right)
        if left == -1 or right == -1 or left - right > 1 or right - left > 1:
            return -1
        return 1 + max(left, right)

File: 110.平衡二叉树.cc
#include <algorithm>
using std::max;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isBalanced(TreeNode *root) {
        if (maxDepth(root) == -1)
            return false;
        return true;
    }

    int maxDepth(TreeNode *root) {
        if (root == nullptr)
            return 0;
        int left = maxDepth(root->left);
        int right = maxDepth(root->right);
        if (left == -1 || right == -1 || left - right > 1 || right - left > 1)
            return -1;
        return 1 + max(left, right);
    }
};
